### PR TITLE
GH-20086: [C++] Cast between fixed size and variable size lists

### LINKED ISSUE #20086: [C++] Cast between fixed size list type and variable size list
Casting from fixed size list to variable size list could be possible, I think, but currently doesn't work:

```python

>>> fixed_size = pa.array([[1, 2], [3, 4]], type=pa.list_(pa.int64(), 2))
>>> fixed_size.cast(pa.list_(pa.int64()))
...
ArrowNotImplementedError: Unsupported cast from fixed_size_list<item: int64>[2] to list using function cast_list
```

And in principle, a cast the other way around could also be possible if it is checked that each list has the correct length.

**Reporter**: [Joris Van den Bossche](https://issues.apache.org/jira/browse/ARROW-15455) / @jorisvandenbossche
#### PRs and other links:
- [GitHub Pull Request #12517](https://github.com/apache/arrow/pull/12517)

<sub>**Note**: *This issue was originally created as [ARROW-15455](https://issues.apache.org/jira/browse/ARROW-15455). Please see the [migration documentation](https://github.com/apache/arrow/issues/14542) for further details.*</sub>

### PR DESCRIPTION
**Pull request by wjones127**
### Rationale for this change

These list types should be able to cast to each other. In our code base, we find users might naively make an list array (for example, with `pa.Table.from_pylist()`) and want it automatically casted to a tensor array (which is based on a fixed size list).

### What changes are included in this PR?

Adds casts from FSL -> (Large)List and (Large)List -> FSL.

### Are these changes tested?

The kernels are tested.

### Are there any user-facing changes?

Just a new kernel.
* Closes: #20086

### CODE REVIEW

**Comment by wjones127** [comment]
Python example:

```python
>>> import pyarrow as pa
>>> arr = pa.array([[1, 2, 3], [4, 5, 6]])
>>> arr.type
ListType(list<item: int64>)
>>> arr.cast(pa.list_(pa.int64(), 3))
<pyarrow.lib.FixedSizeListArray object at 0x132793e80>
[
  [
    1,
    2,
    3
  ],
  [
    4,
    5,
    6
  ]
]
>>> arr.cast(pa.list_(pa.int64(), 3)).cast(pa.list_(pa.int64()))
<pyarrow.lib.ListArray object at 0x132793ee0>
[
  [
    1,
    2,
    3
  ],
  [
    4,
    5,
    6
  ]
]
>>> arr.cast(pa.list_(pa.int64(), 2))
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "pyarrow/array.pxi", line 967, in pyarrow.lib.Array.cast
    return _pc().cast(self, target_type, safe=safe,
  File "/Users/willjones/Documents/arrow/python/pyarrow/compute.py", line 403, in cast
    return call_function("cast", [arr], options, memory_pool)
  File "pyarrow/_compute.pyx", line 572, in pyarrow._compute.call_function
    return func.call(args, options=options, memory_pool=memory_pool,
  File "pyarrow/_compute.pyx", line 367, in pyarrow._compute.Function.call
    result = GetResultValue(
  File "pyarrow/error.pxi", line 144, in pyarrow.lib.pyarrow_internal_check_status
    return check_status(status)
  File "pyarrow/error.pxi", line 100, in pyarrow.lib.check_status
    raise ArrowInvalid(message)
pyarrow.lib.ArrowInvalid: ListType can only be casted to FixedSizeListType if the lists are all the expected size.
/Users/willjones/Documents/arrow/cpp/src/arrow/compute/exec.cc:920  kernel_->exec(kernel_ctx_, input, &output)
/Users/willjones/Documents/arrow/cpp/src/arrow/compute/function.cc:276  executor->Execute(input, &listener)
```

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 1]
```diff
@@ -33,11 +34,24 @@ namespace arrow {
 
 using internal::CopyBitmap;
 
-namespace compute {
-namespace internal {
+namespace compute::internal {
```
Comment (wjones127) [inline cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:37]: This is for a clang-tidy lint.

**Comment by pitrou** [comment]
@js8544 Do you want to review this?

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 2]
```diff
@@ -2326,6 +2326,76 @@ TEST(Cast, FSLToFSLOptionsPassThru) {
...
+  auto options = CastOptions::Safe(list(int16()));
+  CheckCastFails(fsl_int32, options);
+  options.allow_int_overflow = true;
+  CheckCast(fsl_int32, ArrayFromJSON(fixed_size_list(int16(), 1), "[[32689]]"), options);
+
+  // Invalid fixed_size_list cast if inconsistent size
+  ASSERT_RAISES(Invalid, Cast(ArrayFromJSON(list(int32()), "[[0, 1, 2], null, [3, 4]]"),
+                              CastOptions::Safe(fixed_size_list(int32(), 3))))
+      << "Size of FixedList is not the same.";
+}
```
Comment (js8544) [inline cpp/src/arrow/compute/kernels/scalar_cast_test.cc:2397]: Should we also have tests for:
1. Inputs with `null`s not at the end.
2. Inputs with manually constructed offsets where `null` elements have non-zero value length. `ArrayFromJSON` always has `offset[i+1]==offset[i]` for null elements.

**Review by js8544 — CHANGES_REQUESTED** [review]

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 3]
```diff
@@ -145,6 +153,135 @@ void AddListCast(CastFunction* func) {
...
+    auto* offsets = out_array->GetMutableValues<dest_offset_type>(1);
+    dest_offset_type offset = 0;
+    for (int64_t i = 0; i <= batch.length; ++i) {
+      offsets[i] = offset;
+      offset += list_size;
+    }
+
+    // Handle values
+    std::shared_ptr<ArrayData> values = in_array.child_data[0].ToArrayData();
+    if (in_array.offset > 0 || (in_array.length * list_size < batch.length)) {
```
[the remark this answers is not in the document]
Comment (wjones127) [reply cpp/src/arrow/compute/kernels/scalar_cast_nested.cc:186]: Actually, I think the second condition might be unnecessary. The offsets will already determine the end of the array, so we shouldn't need to slice the values if there isn't an offset.

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc (hunk 2 above)
Comment (wjones127) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc:2409]: Added an example using `MaskArrayWithNullsAt` and added some nulls in the first examples.

File: cpp/src/arrow/compute/kernels/scalar_cast_test.cc [hunk 4]
```diff
@@ -2326,6 +2326,76 @@ TEST(Cast, FSLToFSLOptionsPassThru) {
...
+  auto fsl_int32 = ArrayFromJSON(fixed_size_list(int32(), 1), "[[87654321]]");
+  auto options = CastOptions::Safe(list(int16()));
+  CheckCastFails(fsl_int32, options);
+  options.allow_int_overflow = true;
+  CheckCast(fsl_int32, ArrayFromJSON(fixed_size_list(int16(), 1), "[[32689]]"), options);
+
+  // Invalid fixed_size_list cast if inconsistent size
+  ASSERT_RAISES(Invalid, Cast(ArrayFromJSON(list(int32()), "[[0, 1, 2], null, [3, 4]]"),
+                              CastOptions::Safe(fixed_size_list(int32(), 3))))
+      << "Size of FixedList is not the same.";
```
[the remark this answers is not in the document]
Comment (wjones127) [reply cpp/src/arrow/compute/kernels/scalar_cast_test.cc]: Removed that.

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 5]
```diff
@@ -145,6 +153,135 @@ void AddListCast(CastFunction* func) {
...
+
+    // Handle values
+    std::shared_ptr<ArrayData> values = in_array.child_data[0].ToArrayData();
+    ARROW_ASSIGN_OR_RAISE(Datum cast_values_datum,
+                          Cast(values, child_type, options, ctx->exec_context()));
+
+    DCHECK(cast_values_datum.is_array());
+    std::shared_ptr<ArrayData> cast_values = cast_values_datum.array();
+
+    if (in_array.GetNullCount() > 0 && in_array.length > 0) {
```
[the remark this answers is not in the document]
Comment (wjones127) [reply cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: That makes sense. Removed.

File: cpp/src/arrow/compute/kernels/scalar_cast_nested.cc [hunk 6]
```diff
@@ -145,6 +153,135 @@ void AddListCast(CastFunction* func) {
...
+    } else {
+      // Don't need to check null slots if there are no nulls
+      for (int64_t i = 0; i <= batch.length; ++i) {
+        if (offsets[i] != expected_offset) {
+          return Status::Invalid("ListType can only be casted to FixedSizeListType ",
+                                 "if the lists are all the expected size.");
+        }
+        expected_offset += list_size;
+      }
+    }
```
[the remark this answers is not in the document]
Comment (wjones127) [reply cpp/src/arrow/compute/kernels/scalar_cast_nested.cc]: You are right the null part wasn't handled. Switched to just checking the ends.

**Review by js8544 — CHANGES_REQUESTED** [review]

**Review by js8544 — APPROVED** [review]

**Comment by jorisvandenbossche** [comment]
Thanks @wjones127!

**Comment by felipecrv** [comment]
One of the tasks in completing list-view support [1] is making all these list and list-view types "castable" between themselves. I even started writing some of the code for this (not pushed to any PR at the moment).

[1] #37249

**Pull request merged**